apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/RowAsyncEnumerator.cs

```diff
@@ -0,0 +1,298 @@
+                _stream.GrpcCall.GetStatus().StatusCode != StatusCode.Cancelled)
+            {
+                // TODO: Is there something we could/should do here? Stream prematurely closed
+                //       with a row in progress
+            }
+
+            return false;
+
+            IEnumerator<Row> MergeResponseChunks(ReadRowsResponse response)
+            {
+                foreach (var chunk in response.Chunks)
+                {
+                    if (chunk.ResetRow)
+                    {
+                        Reset();
+                        continue;
+                    }
+
+                    _rowMergeState = _rowMergeState.HandleChunk(this, chunk);
+
+                    if (chunk.CommitRow)
+                    {
+                        _rowMergeState = _rowMergeState.CommitRow(this);
+
+                        // TODO: Capture response.LastScannedRowKey here for smart retries
+                        yield return _currentCell.Row;
+                    }
+                }
+            }
+        }
+
+        public bool IsRowInProgress => _rowMergeState != NewRow.Instance;
+
+        private void AddCompletedCellToRow()
+        {
+            var cell = new Cell
+            {
+                Value = _currentCell.Value,
+                TimestampMicros = _currentCell.Timestamp,
+                Labels = { _currentCell.Labels }
+            };
+            Assert(_currentCell.Column != null, "NewCell has no qualifier");
+            _currentCell.Column.Cells.Add(cell);
+        }
+
+        private void Assert(bool condition, string message)
+        {
+            if (!condition)
+            {
+                Reset();
+
+                // TODO: Is there something better we should throw here?
+                if (message != null)
+                {
+                    throw new InvalidOperationException(message);
+                }
+                else
+                {
+                    throw new InvalidOperationException();
+                }
+            }
+        }
+
+        private void Reset() => _currentCell = new CellInfo();
+
+        private struct CellInfo
+        {
+            public Column Column { get; set; }
+            public Family Family { get; set; }
+            public IEnumerable<string> Labels { get; set; }
+            public Row Row { get; set; }
+            public long Timestamp { get; set; }
+            public ByteString Value { get; set; }
+            public int ValueSizeExpected { get; set; }
+            public int ValueSizeRemaining { get; set; }
+        }
+
+        private abstract class RowMergeState
+        {
+            public abstract RowMergeState HandleChunk(RowAsyncEnumerator owner, CellChunk chunk);
+
+            public virtual RowMergeState CommitRow(RowAsyncEnumerator owner)
+            {
+                owner.Assert(false, null);
+                return this;
+            }
+        }
+
+        private sealed class NewRow : RowMergeState
```

Comment:
Comments for these states would be really useful :)

Reply:
Oops, meant to do that on a final pass and forgot. Done.